pkg/template/processor_test.go

```diff
@@ -0,0 +1,807 @@
+		objs, err := p.Process(tmpl, values)
+		require.NoError(t, err)
+		err = p.Apply(objs)
+		require.NoError(t, err)
+		assertRoleBindingExists(t, cl, user)
+
+		// when rolebinding changes
+		tmpl, err = decodeTemplate(decoder,
+			createTemplate(withObjects(namespace, rolebindingWithExtraUser), withParams(usernameParam, commitParam)))
+		require.NoError(t, err)
+		objs, err = p.Process(tmpl, values)
+		require.NoError(t, err)
+		err = p.Apply(objs)
+
+		// then
+		require.NoError(t, err)
+		binding := assertRoleBindingExists(t, cl, user)
+		require.Len(t, binding.Subjects, 2)
+		assert.Equal(t, "User", binding.Subjects[0].Kind)
+		assert.Equal(t, user, binding.Subjects[0].Name)
+		assert.Equal(t, "User", binding.Subjects[1].Kind)
+		assert.Equal(t, "extraUser", binding.Subjects[1].Name)
+	})
+
+	t.Run("failures", func(t *testing.T) {
+
+		t.Run("should fail to create template object", func(t *testing.T) {
+			// given
+			cl := test.NewFakeClient(t)
+			p := template.NewProcessor(cl, s)
+			cl.MockCreate = func(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error {
+				return errors.New("failed to create resource")
+			}
+			tmpl, err := decodeTemplate(decoder,
+				createTemplate(withObjects(rolebinding), withParams(usernameParam, commitParam)))
+			require.NoError(t, err)
+
+			// when
+			objs, err := p.Process(tmpl, values)
+			require.NoError(t, err)
+			err = p.Apply(objs)
+
+			// then
+			require.Error(t, err)
+		})
+
+		t.Run("should fail to update template object", func(t *testing.T) {
+			// given
+			cl := test.NewFakeClient(t)
+			p := template.NewProcessor(cl, s)
+			cl.MockUpdate = func(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
+				return errors.New("failed to update resource")
+			}
+			tmpl, err := decodeTemplate(decoder,
+				createTemplate(withObjects(rolebinding), withParams(usernameParam, commitParam)))
+			require.NoError(t, err)
+			objs, err := p.Process(tmpl, values)
+			require.NoError(t, err)
+			err = p.Apply(objs)
+			require.NoError(t, err)
+
+			// when
+			tmpl, err = decodeTemplate(decoder,
+				createTemplate(withObjects(rolebindingWithExtraUser), withParams(usernameParam, commitParam)))
+			require.NoError(t, err)
+			objs, err = p.Process(tmpl, values)
+			require.NoError(t, err)
+			err = p.Apply(objs)
+
+			// then
+			assert.Error(t, err)
+		})
+	})
+
+	t.Run("should create with extra labels and ownerref", func(t *testing.T) {
+
+		// given
+		values := map[string]string{
+			"USERNAME": user,
+			"COMMIT":   commit,
+		}
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+		tmpl, err := decodeTemplate(decoder,
+			createTemplate(withObjects(namespace, rolebinding), withParams(usernameParam, commitParam)))
+		require.NoError(t, err)
+		objs, err := p.Process(tmpl, values)
+		require.NoError(t, err)
+
+		// when adding labels and an owner reference
+		obj := objs[0]
+		nsObj, ok := obj.Object.(*unstructured.Unstructured)
```

Comment:
should you cast to `runtime.Unstructured` here?

Reply:
ah, I missed this one - fixed here 521794c but didn't use any `*.Unstructured` at all ;-)